[LSP] - Option  to  truncate LSP diagnostic virtual text on newline.

ckipp01 opened the pull request:
Currently when we have a diagnostic that contains a new line, we simply
strip it out and replace it with a space. For some languages this may
work fine, but for a lot of typed languages that have longer signatures
or types and also multi-line diagnostics, this falls apart pretty quickly.
This takes a different approach and if a new line is detected it will
instead truncate the diagnostic message (only for the virtual text) and
only display the first line making a more succinct message. Then when they
go to view the diagnostic they will continue to see the full message as they
would have before.

Here are a few images to illustrate this. The current behavior looks like this:

_A longer example with a stack trace include, which will never show correctly_
<img width="1680" alt="Screenshot 2021-01-09 at 11 12 35" src="https://user-images.githubusercontent.com/13974112/104089127-80748e80-526c-11eb-8c03-f0d3411631e2.png">

_A shorter example that already causes issues if the screen is split_
<img width="851" alt="Screenshot 2021-01-09 at 11 13 49" src="https://user-images.githubusercontent.com/13974112/104089133-908c6e00-526c-11eb-8b2d-8e4828c18d71.png">

The proposed behavior would look like this:

_Longer example now only shows the first line_
<img width="1128" alt="Screenshot 2021-01-09 at 11 03 10" src="https://user-images.githubusercontent.com/13974112/104089142-ab5ee280-526c-11eb-8015-ccf49d052f5d.png">

_To see the rest of the diagnostic, you'd still be able to view it as before_
<img width="1151" alt="Screenshot 2021-01-09 at 11 03 26" src="https://user-images.githubusercontent.com/13974112/104089158-c5002a00-526c-11eb-9930-cf182048cc16.png">

_And even the shorter examples still get the main point across_
<img width="551" alt="Screenshot 2021-01-09 at 11 08 45" src="https://user-images.githubusercontent.com/13974112/104089164-d77a6380-526c-11eb-996e-60c1e7fce2dd.png">

_And again with the details still there_
<img width="551" alt="Screenshot 2021-01-09 at 11 08 53" src="https://user-images.githubusercontent.com/13974112/104089170-e2cd8f00-526c-11eb-91d2-f066277d3d56.png">

ckipp01 committed:
Truncate LSP diagnostic virtual text on new line.

Currently when we have a diagnostic that contains a new line, we simply
strip it out and replace it with a space. For some languages this may
work fine, but for a lot of typed languages that have longer signatures
or types and also multiline diagnostics, this falls apart pretty quickly.
This takes a different approach and if a new line is detected it will
instead trunacate the diagnostic message (only for the virtual text) and
only display the first line making a more succinct message. Then when they
go to view the diagnostic they will continue to see the full message as they
would have before.

teto commented:
if you have a wide screen you may prefer to have all content ?  it would be nice to share that function in the wiki though so people can use it. or modify get_virtual_text_chunks_for_line to accept a function that processes the message so that it is more easily overridable

ckipp01 commented:
> if you have a wide screen you may prefer to have all content ? it would be nice to share that function in the wiki though so people can use it. or modify get_virtual_text_chunks_for_line to accept a function that processes the message so that it is more easily overridable

I guess maybe a better question would be what is the main purpose of displaying the error message with virtual text? Is it truly meant to show the entire erorr message, or simply notify the user that there is an error on that line, and give a preview of what it may be. Mainly since there will be situations, especially with stack traces, where it's impossible to show the whole thing. It would just never work. So imo this seems to be a much cleaner solution. I'd say this is more in line with what diagnostics often show in other editors, where normally it's just an underline signifying it.

>  or modify get_virtual_text_chunks_for_line to accept a function that processes the message so that it is more easily overridable

I guess I'm willing to do this, but I'm a bit skeptical that it would be used that often. What do you think?

ckipp01 commented:
@mfussenegger I'm curious if you ever hit on this with Java.

mfussenegger commented:
> @mfussenegger I'm curious if you ever hit on this with Java.

I can't really tell on top of my head. I've virtual text for diagnostic disabled because I felt like it's too much noise :)


Regarding the proposed changes. How about adding a new option to the `opts` table of `set_virtual_text` and `get_virtual_text_chunks_for_line` so that users can enable/disable the trimming? See `:help vim.lsp.diagnostic.on_publish_diagnostics` - users can customize the options using `vim.lsp.with`

ckipp01 commented:
> Regarding the proposed changes. How about adding a new option to the `opts` table of `set_virtual_text` and `get_virtual_text_chunks_for_line` so that users can enable/disable the trimming? See `:help vim.lsp.diagnostic.on_publish_diagnostics` - users can customize the options using `vim.lsp.with`

Ah, yea that's a good idea. I'll look into doing this.

ckipp01 committed:
Merge remote-tracking branch 'upstream/master'

ckipp01 committed:
Add truncated in as a opt.

This will allow for a user to do the following

```lua
vim.lsp.with(vim.lsp.diagnostic.on_publish_diagnostics,
                                                {virtual_text = {truncated = true}})
```

Which will then truncate the virtual text on newline.

ckipp01 commented:
Alright, I followed your suggest @mfussenegger and add in a `truncated` field to the `opt`. So now this is able to be used like so:

```lua
vim.lsp.with(vim.lsp.diagnostic.on_publish_diagnostics, {virtual_text = {truncated = true}})
                                               
```
So if it's truncated, it will truncate on the `\n`, and if not, the behavior remains the same.

@teto this should be ready to review now.

ckipp01 commented:
@mjlbach pinging you on here as well in case you had any input.

ckipp01 commented:
@teto or @tjdevries, and thoughts on this?

tjdevries commented on runtime/lua/vim/lsp/diagnostic.lua:
```diff
@@ -782,10 +783,23 @@ function M.get_virtual_text_chunks_for_line(bufnr, line, line_diags, opts)
   -- TODO(tjdevries): Allow different servers to be shown first somehow?
   -- TODO(tjdevries): Display server name associated with these?
   if last.message then
+    local stripped_msg = last.message:gsub("\r", "")
+    local newline_index, _ = string.find(stripped_msg, "\n")
+    local diagnostic_msg
+    if newline_index then
+      if opts.truncated then
+        diagnostic_msg = string.sub(stripped_msg, 1, newline_index - 1)
+      else
+        diagnostic_msg = stripped_msg:gsub("\n", " ")
+      end
+    else
+      diagnostic_msg = stripped_msg
+    end
+
```
maybe a better solution would be to expose a option that is a function that takes in the `function(client, message) -> message`. This way people can do whatever they want with the message and we don't have a multiplication of all the possible configurations someone could have.

What do you think about doing that?

(Also, I'd be fine with changing our  default to only show the first line and a `...` if there is more text on other lines or something, since otherwise the virtual text can be way too big).

ckipp01 replied:
> maybe a better solution would be to expose a option that is a function that takes in the function(client, message) -> message. This way people can do whatever they want with the message and we don't have a multiplication of all the possible configurations someone could have.

In some way I can see the benefit of that for sure, but the negative side of this would be that there still isn't an easy way to have a sane default. It's either, you get everything (even though they may be insanely long), or you write your own function to deal with it. The reason I liked this approach was it was just a simple `truncated = true` that would be required.

I'd be way more inclined to do your second suggestion:

> (Also, I'd be fine with changing our default to only show the first line and a ... if there is more text on other lines or something, since otherwise the virtual text can be way too big).

Since I think this is way more sane. If you're ok with doing this, I'm ok with making the change.

tjdevries replied:
Ah, sorry, I was saying we could do _both_ of those. The default `function(client, message) -> message` would return the truncated string. Or perhaps it would be better to pass as a table, so we can change the arguments later. `{ client = client, message = message, ... }` might be easier to extend later.

I think we can do it both ways (and this would increase the configurability as time goes on).

tjdevries replied:
(the function + changing the default means we don't have to provide `truncated` as a new configuration option though)

ckipp01 replied:
ah ok, gotcha gotcha. Sounds good. I'll take a stab at this then and offer up both 👍🏼

ckipp01 replied:
Quick question on this. Just to make sure I'm understanding, this function would still be exposed as an option under `virtual_text` though right? So if someone was going to use it, it'd be 

```lua
local my_function_to_handle_diagnotsic_virtual_text = function({client, message}) ... end

vim.lsp.with(vim.lsp.diagnostic.on_publish_diagnostics, {
  virtual_text = {
    display = my_function_to_handle_diagnotsic_virtual_text -- is display a good name?
  }
})
```
I just want to make sure I'm following how you envision this.

tjdevries replied:
Yeah, exactly that style. 

I'd say something like `get_message` would be better -- something with a verb gives a hint that it's a function.

ckipp01 commented:
I won't have time to work on this for a bit, so I want to close it to potentially unblock anyone that wants to take a stab at this. I did create an issue for it here: https://github.com/neovim/neovim/issues/13860
